### PR TITLE
Resolve Web UI Beta Testing Bugs

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

**Which issues(s) does this PR fix?**

- Includes the validator status in the list balances response
- Adds a default wallet response endpoint
- Logs in a user if they try to signup and already have an account
- Ensures we have stronger password validation in place
- Use a package level cache for keys in the keymanagers
- Remove SendDepositTx functionality in derived keymanager
- Only send a wallet over the WalletInitializedFeed if we have validating public keys

**Other notes for review**

Diff large due to mocks

### CODE REVIEW

File: validator/rpc/auth.go [hunk 1]
```diff
@@ -22,30 +22,34 @@ import (
...
-	tokenExpiryLength = 20 * time.Minute
+	tokenExpiryLength = time.Hour
 	hashCost          = 8
 )
 
 // Signup to authenticate access to the validator RPC API using bcrypt and
 // a sufficiently strong password check.
 func (s *Server) Signup(ctx context.Context, req *pb.AuthRequest) (*pb.AuthResponse, error) {
+	walletDir := s.walletDir
+	if req.WalletDir != "" {
```
Comment (rkapka) [inline validator/rpc/auth.go:33]: Maybe trim the value before comparison to disallow dirs containing whitespace only.

File: validator/rpc/auth.go [hunk 2]
```diff
@@ -56,19 +60,29 @@ func (s *Server) Signup(ctx context.Context, req *pb.AuthRequest) (*pb.AuthRespo
...
-		hashedPassword, err := fileutil.ReadFileAsBytes(hashedPasswordPath)
-		if err != nil {
-			return nil, status.Error(codes.Internal, "Could not retrieve hashed password from disk")
-		}
-		// Compare the stored hashed password, with the hashed version of the password that was received.
-		if err := bcrypt.CompareHashAndPassword(hashedPassword, []byte(req.Password)); err != nil {
-			return nil, status.Error(codes.Unauthenticated, "Incorrect password")
-		}
+	walletDir := s.walletDir
+	if req.WalletDir != "" {
```
Comment (rkapka) [inline validator/rpc/auth.go:64]: Same comment about trimming.

File: validator/rpc/wallet.go [hunk 3]
```diff
@@ -47,35 +47,50 @@ func (s *Server) HasWallet(ctx context.Context, _ *ptypes.Empty) (*pb.HasWalletR
...
+	}, nil
+}
+
 // CreateWallet via an API request, allowing a user to save a new
 // derived, direct, or remote wallet.
 func (s *Server) CreateWallet(ctx context.Context, req *pb.CreateWalletRequest) (*pb.CreateWalletResponse, error) {
-	// Currently defaultWalletPath is used as the wallet directory and req's WalletPath is ignored for simplicity
-	exists, err := wallet.Exists(defaultWalletPath)
+	walletDir := s.walletDir
+	if req.WalletPath != "" {
```
Comment (rkapka) [inline validator/rpc/wallet.go:61]: Same comment about trimming.

File: validator/rpc/wallet.go [hunk 4]
```diff
@@ -252,6 +262,20 @@ func (s *Server) ChangePassword(ctx context.Context, req *pb.ChangePasswordReque
...
+		return nil, status.Error(codes.FailedPrecondition, "Could not compare password from disk")
+	}
+	hashedPassword, err := fileutil.ReadFileAsBytes(hashedPasswordPath)
+	if err != nil {
+		return nil, status.Error(codes.FailedPrecondition, "Could not retrieve hashed password from disk")
+	}
+	if err := bcrypt.CompareHashAndPassword(hashedPassword, []byte(req.CurrentPassword)); err != nil {
+		return nil, status.Error(codes.Unauthenticated, "Incorrect wallet password")
+	}
 	if req.Password == "" {
```
Comment (rkapka) [inline validator/rpc/wallet.go:279]: Same comment about trimming.

File: validator/rpc/wallet.go [hunk 5]
```diff
@@ -252,6 +262,20 @@ func (s *Server) ChangePassword(ctx context.Context, req *pb.ChangePasswordReque
 	if !valid {
 		return nil, status.Errorf(codes.FailedPrecondition, invalidWalletMsg)
 	}
+	if req.CurrentPassword == "" {
+		return nil, status.Error(codes.InvalidArgument, "Current wallet password cannot be empty")
+	}
```
Comment (rkapka) [inline validator/rpc/wallet.go:267]: Nitpick alert

I am not sure it makes much sense to do any additional checks except comparing the current password with the one from disk. We disallow empty passwords in the first place anyway.

File: beacon-chain/rpc/beacon/validators.go [hunk 6]
```diff
@@ -838,3 +846,26 @@ func validatorHasExited(validator *ethpb.Validator, currentEpoch uint64) bool {
 	}
 	return true
 }
+
+func validatorStatus(validator *ethpb.Validator, epoch uint64) ethpb.ValidatorStatus {
```
Comment (rauljordan) [inline beacon-chain/rpc/beacon/validators.go:850]: TODO: add a test or abstract this somewhere else

File: proto/validator/accounts/v2/web_api.proto [hunk 7]
```diff
@@ -115,24 +120,24 @@ message CreateWalletRequest {
...
 }
 
 message CreateWalletResponse {
-   WalletResponse wallet = 1;
-   DepositDataResponse accounts_created = 2;
+    WalletResponse wallet = 1;
+    DepositDataResponse accounts_created = 2;
+}
+
+message DefaultWalletResponse {
```
Comment (prestonvanloon) [inline proto/validator/accounts/v2/web_api.proto:139]: Is a new message required? This looks like WalletResponse.

File: proto/validator/accounts/v2/web_api.proto [hunk 8]
```diff
@@ -18,6 +18,11 @@ service Wallet {
             body: "*"
         };
     }
+    rpc DefaultWalletPath(google.protobuf.Empty) returns (DefaultWalletResponse) {
```
Comment (prestonvanloon) [inline proto/validator/accounts/v2/web_api.proto:21]:
```suggestion
    rpc DefaultWallet(google.protobuf.Empty) returns (WalletResponse) {
```
I think the "Path" part doesn't align with the path. Also, can you use `WalletResponse`? 
That seems to make more sense to me.

File: proto/validator/accounts/v2/web_api.proto [hunk 9]
```diff
@@ -18,6 +18,11 @@ service Wallet {
...
+        option (google.api.http) = {
+            get: "/v2/validator/wallet/default",
```
Comment (prestonvanloon) [inline proto/validator/accounts/v2/web_api.proto:23]:
```suggestion
            get: "/v2/validator/wallet/config/default",
```
Recommend: add config to the path for clarity. 
I was thinking that `/v2/validator/wallet` would return a default wallet, so the distinction wasn't very clear between those two API routes.

File: validator/web/handler_test.go [hunk 10]
```diff
@@ -37,7 +37,7 @@ func TestWebHandler(t *testing.T) {
 			name:            "favicon.ico",
 			requestURI:      "/favicon.ico",
 			wantStatus:      200,
-			wantContentType: "image/vnd.microsoft.icon",
+			wantContentType: "image/x-icon",
```
Comment (prestonvanloon) [inline validator/web/handler_test.go:40]: Interesting... was this failing for you?

**Review by prestonvanloon — DISMISSED** [review]
LGTM

Discussed offline that API changes / refactoring can be done in a follow up PR.
If you create a tracking issue, please link to this PR. Thanks!

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**